feat(search): handle search indexing coordination among nodes

kuangp opened the pull request:
Signed-off-by: Phil Kuang <pkuang@factset.com>

## Hey, I just made a Pull Request!

Fixes #7784

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

iamEAP reviewed (commented):
This is really fantastic stuff, @kuangp!  Thanks for digging in.  Implementation looks 👌 .

Might be good to update a few areas of documentation along with this:

- Code snippets and some text on [the getting started](https://backstage.io/docs/features/search/getting-started#backend)
- Some detail on coordination in the scheduler section of the [concepts page](https://backstage.io/docs/features/search/concepts#the-scheduler)

freben commented:
Hey! Sorry to muddy the PR with unrelated discussion, but this thing made me think - especially in relation to the lunr question by @iamEAP - maybe the scheduling shouldn't be driven from inside the index builder at all? I see it copying some parameters that actually are just a passthrough into the scheduling.

Taking some inspiration out of the [EntityProvider + Connection](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/providers/types.ts) duality, would it make sense if the index builder offered to the ability to collators to push changes over some "connection like" object?

The installation would now be more like either

```ts
const c = MyCollator.fromConfig(...);
indexBuilder.addCollator(c);
await taskScheduler.scheduleTask(..., () => c.run());
```

Or maybe better,

```ts
indexBuilder.addCollator(
  MyCollator
    .fromConfig()
    .schedule(taskScheduler, { /* some of the options directly out of TaskScheduler */ })
);
```

Then the index builder would know nothing about scheduling, and instead it just calls `async start()` on each collator at startup and then every now and then receives "events" for large updates to push through.

And the lunr one would have a different `schedule` method with other arguments than the distributed ones. That method would not be part of the interface, but rather the concrete class as a convention, and just store the intent of what to do once the `start` happens, and then `return this` at the end. The lunr one would always schedule locally.

Alright, done waiting for builds, just food for thought :)

iamEAP commented:
> Hey! Sorry to muddy the PR with unrelated discussion, but this thing made me think - especially in relation to the lunr question by @iamEAP - maybe the scheduling shouldn't be driven from inside the index builder at all?

Interesting thoughts!  And brings up maybe more of a foundational question from me behind the intention of the task scheduler @freben: is there any particular intention behind where task scheduler code is invoked?  Was the intention to keep task scheduler invocation within a `plugins/foo.ts` execution context?  Or is it fine for a task scheduler to be _injected into_ plugin code and then executed within the plugin?

freben commented:
It's fine to inject! This was more about clarifying responsibilities in this particular instance.

iamEAP commented:
Hmm...  Maybe a little bit of mix-up between search engines and collators in your thinking, @freben?  Or maybe I'm just having a hard time parsing your pseudo-code.  I'm not sure it's appropriate for collators to know anything about scheduling.  Nor search engines either, really.  My mental model:

- **Collators**: _"All I know is how to retrieve documents."_
- **Decorators**: _"All I know is how to modify and filter documents."_
- **Search Engine**: _"All I know is how to store and query sets of documents."_
- **Index Builder**: _"All I know is how to wire up collators, decorators, and search engines into a pipeline."_

Now that the first three are actually all just streams, and the fourth more or less just pipes them together (sorry @kuangp, gonna need a rebase w/minor changes 😬 ) I'm not sure the "connection object" comparison with entity providers is quite the right mental model.

---

Perhaps the place to address the "lunr in prod" warning/error messaging is in the index builder's constructor?  We could (additionally) explicitly pass in the database manager (which would make it clearer that such a dependency exists--it's a little more hidden, when we only rely on the `taskScheduler` dependency), and check "is lunr search engine? is pg (or just not sqlite) knex client? then warn about possible misconfiguration"

freben commented:
Oh for sure, in a streaming scenario any comments regarding scheduling may not be relevant at all and can be ignored :) Sorry also if I mixed some terms.

If scheduling is still of interest, note that we just merged this one https://github.com/backstage/backstage/pull/10018/files#diff-9e7abe371d1c9ea77c6222428a56dba520d8a0cbf0d27c8d4c22e9a3fb62b6c8R81 which may be a pattern of interest; it just passes a `TaskRunner` into the constructor, e.g.

```ts
LdapOrgEntityProvider.fromConfig(config, {
  schedule: scheduler.createScheduledTaskRunner({
    frequency: Duration.fromObject({ minutes: 30 }),
    // ..
```

this detaches the class entirely from what scheduling "means", it's just a runner callback basically after that

kuangp commented:
thanks @iamEAP @freben for your feedback and apologies for the delay here (been OOO and busy recently). 

I updated my change based on your suggestions to shift all scheduling logic to the consumer by having them be responsible for supplying a scheduled task runner (following the LDAP pattern). This also gives consumers more control over configuring task scheduling based on their use-case eg. using the `TaskScheduler` for distributed setups or implementing a "local" task runner when using `lunr`. I also added notes in the docs and changeset regarding the "local" setup to explicitly call out this use-case. 

Let me know what you think!

iamEAP commented on .changeset/ninety-eggs-argue.md:
````diff
@@ -0,0 +1,79 @@
+/* ... */
+```
+
+NOTE: For scenarios where the `lunr` search engine is used in a multi-node configuration, a non-distributed `TaskRunner` like the following should be implemented to ensure consistency across nodes:
````
Does the default task runner work on a sqlite connection?  If so, maybe a simpler recommendation would be to have integrators configure the search plugin to use sqlite as its database instead. [e.g.](https://backstage.io/docs/tutorials/configuring-plugin-databases#postgresql-and-sqlite-3)

iamEAP replied:
And I still wonder if it's possible to detect/warn about this somewhere in the index builder.  ...But I wouldn't call that a blocker.

kuangp replied:
AFAIK, the default task runner _should_ be DB agnostic. @freben any limitations you know of?

iamEAP replied:
It certainly works locally for me, at least!  Could simplify things quite a bit / de-emphasize the use-case with instructions like...

````md
NOTE: Although the following scenario is not recommended, if you run the in-memory `lunr` search engine in a multi-node configuration, you must also configure the `search` plugin to use an in-memory database in order to ensure index consistency across nodes:

```diff
backend:
  database:
    client: pg
    connection: 'postgresql://foo:bar@some.example-pg-instance.tld:5432'
+    plugin:
+      search:
+        client: better-sqlite3
+        connection: ':memory:'
```
````

iamEAP replied:
I can fix the documentation in a follow-up PR (I wanna spend a little bit of time to updating the architecture diagram to better highlight this anyway).

iamEAP commented on docs/features/search/getting-started.md:
```diff
@@ -163,9 +164,14 @@ export default async function createPlugin(
     searchEngine,
   });
 
+  const catalogIndexSchedule = env.scheduler.createScheduledTaskRunner({
```
Trying to grok the implications of collators sharing a schedule (seems like it's totally fine because the coordination is based on the task ID, which is derived from the collator type?)...

Wondering if it would clarify that if we encouraged schedule naming to reflect the frequency, rather than the collator?  e.g. `const every10MinutesSchedule = ...` or similar

iamEAP commented on plugins/search-backend-node/src/Scheduler.ts:
```diff
@@ -14,52 +14,67 @@
       );
     }
-    this.schedule.push({ task, interval });
+    this.schedule[id] = { task, scheduledRunner };
```
Super minor: I wonder if it's worth checking if the key already exists and throwing instead of clobbering -- I guess previously we allowed multiple instances of a collator to be scheduled and run, though it's probably not a valid use-case.

kuangp commented:
@iamEAP made some small tweaks based on your comments, let me know what you think!

kuangp commented on .changeset/ninety-eggs-argue.md:
```diff
@@ -0,0 +1,80 @@
+
++  const schedule = env.scheduler.createScheduledTaskRunner({
++    frequency: Duration.fromObject({ seconds: 600 }),
++    timeout: Duration.fromObject({ seconds: 900 }),
```
Good idea! done!

freben reviewed (approved):
Looks good!

kuangp commented on docs/features/search/getting-started.md:
```diff
@@ -161,9 +162,15 @@ export default async function createPlugin(
   });
 
+  const every10MinutesSchedule = env.scheduler.createScheduledTaskRunner({
+    frequency: Duration.fromObject({ seconds: 600 }),
```
you mean you don't only think about times in seconds/milliseconds 😄

iamEAP reviewed (approved):
💯 💯 💯   Thanks again @kuangp!
